kuma/wiki/tests/test_content.py

```diff
@@ -912,97 +836,120 @@ def test_bleach_filter_invalid_protocol(self):
-                <li><a class="foobar" name="quux">A lack of href should not cause a problem.</a></li>
-                <li><a>In fact, a "link" with no attributes should be no problem as well.</a></li>
-                <a href="%(tag_url)s">Tag link</a>
-                <a href="%(feed_url)s">Feed link</a>
-                <a href="%(templates_url)s">Templates link</a>
-                <a href="/en-US/docs/DOM/stylesheet">Case sensitive 1</a>
-                <a href="/en-US/docs/DOM/Stylesheet">Case sensitive 1</a>
-                <a href="/en-US/docs/DOM/StyleSheet">Case sensitive 1</a>
-                <a href="/en-us/docs/dom/StyleSheet">Case sensitive 1</a>
-                <a class="new" href="/en-US/docs/dom/Styles">For good measure</a>
-        """ % vars
-
-        # Split the markup into lines, to better see failures
-        doc_lines = doc_src.strip().split("\n")
-        expected_lines = expected.strip().split("\n")
-        for idx in range(0, len(doc_lines)):
-            doc_line = doc_lines[idx]
-            expected_line = expected_lines[idx]
-            result_line = (kuma.wiki.content.parse(doc_line)
-                                            .annotateLinks(
-                                                base_url=vars['base_url'])
-                                            .serialize())
-            self.assertHTMLEqual(normalize_html(expected_line), normalize_html(result_line))
+def test_annotate_links_encoded_utf8(db):
+    """Encoded UTF8 characters in links are decoded."""
+    Document.objects.create(locale='fr', slug=u'CSS/Héritage',
+                            title=u'Héritée')
+    html = normalize_html(
+        u'<li><a href="/fr/docs/CSS/H%c3%a9ritage">Héritée</a></li>')
+    actual_raw = parse(html).annotateLinks(base_url=AL_BASE_URL).serialize()
+    assert normalize_html(actual_raw) == html
+
+
+@pytest.mark.parametrize('has_class', ('hasClass', 'noClass'))
+@pytest.mark.parametrize('anchor', ('withAnchor', 'noAnchor'))
+@pytest.mark.parametrize('full_url', ('fullURL', 'pathOnly'))
+def test_annotate_links_existing_doc(root_doc, anchor, full_url, has_class):
+    """Links to existing docs are unmodified."""
+    if full_url == 'fullURL':
+        url = root_doc.get_full_url()
+        assert url.startswith(AL_BASE_URL)
+    else:
+        url = root_doc.get_absolute_url()
+    if anchor == 'withAnchor':
+        url += "#anchor"
+    if has_class == 'hasClass':
+        link_class = ' class="extra"'
+    else:
+        link_class = ''
+    html = normalize_html('<li><a %s href="%s"></li>' % (link_class, url))
+    actual_raw = parse(html).annotateLinks(base_url=AL_BASE_URL).serialize()
+    assert normalize_html(actual_raw) == html
+
+
+@pytest.mark.parametrize('has_class', ('hasClass', 'noClass'))
+@pytest.mark.parametrize('anchor', ('withAnchor', 'noAnchor'))
+@pytest.mark.parametrize('full_url', ('fullURL', 'pathOnly'))
```

Comment:
Nice. This is when `parametrize` really shines, handling all of the combinations.